### PR TITLE
feat: pluggable `local/dev/k8s do <job>` commands

### PR DESCRIPTION
**Pull request by regisb**
We introduce a new filter to implement custom commands in arbitrary containers.
It becomes easy to write convenient ad-hoc commands that users will
then be able to run either on Kubernetes or locally using a documented CLI.

Pluggable jobs are declared as Click commands and are responsible for
parsing their own arguments. See the new CLI_DO_COMMANDS filter.

Close https://github.com/overhangio/2u-tutor-adoption/issues/75

### CODE REVIEW

**Comment by regisb** [comment]
@kdmccormick I'm curious to have your thoughts on this.

**Comment by kdmccormick** [comment]
Nice, I'll review shortly @regisb .

File: tutor/commands/jobs.py
Comment (kdmccormick) [inline tutor/commands/jobs.py:62]: Hm, in the interest of reducing the total number of extension points plugin authors need to consider, I was kinda hoping that we could completely implement initialization within the `CLI_DO_COMMANDS` framework, allowing us to deprecate `COMMANDS_INIT` and `COMMANDS_PRE_INIT` and implement `do init` as an ordinary job with no calls to internal Tutor functions.

I understand why it is not currently this way: the `do` framework is designed to allow plugins to *add new* jobs, not to *extend existing* jobs. I don't really have any suggestions here yet... I'll think about it over the weekend.

File: tutor/commands/jobs.py
Comment (kdmccormick) [inline tutor/commands/jobs.py:37]: Since all the updates we've made to the Plugins API, the name and structure of the `hooks/` template directory has stopped making sense.

What about, instead, having a single, flat `init/` directory?

```suggestion
        hooks.Filters.COMMANDS_INIT.add_item(("mysql", ("init", "mysql")))
    with hooks.Contexts.APP("lms").enter():
        hooks.Filters.COMMANDS_INIT.add_item(("lms", ("init", "lms")))
    with hooks.Contexts.APP("cms").enter():
        hooks.Filters.COMMANDS_INIT.add_item(("cms", ("init", "cms")))
```
(I'd also support holding off on these sort of refactorings until the big tutorcore extraction.)

File: tutor/commands/k8s.py
Comment (kdmccormick) [inline tutor/commands/k8s.py:553]: I found that custom commands cannot be added, because this line is run before plugin modules are imported. I was able to fix it by moving this line into a `Actions.PLUGINS_LOADED` callback.

File: tutor/commands/compose.py
Comment (kdmccormick) [inline tutor/commands/compose.py:304]: I think supporting `--mount` would make `do` much more powerful. For example, if we incorporated quickdev, it'd be great to be able to write something like:
```
tutor dev do --mount=../xblock-drag-and-drop-v2 pip-install-mounts
```
And if we didn't incorporate quickdev, then we'll still want some sort of task to job to make [setting up a dev env easier](https://docs.tutor.overhang.io/dev.html#setting-up-a-development-environment-for-edx-platform), like:
```
tutor dev do --mount=../edx-platform edx-platform-setup
```

File: CHANGELOG-nightly.md
Comment (kdmccormick) [inline CHANGELOG-nightly.md:13]: I'm not sure I understand this. Launch/quickstart supports mounting, so why wouldn't init?

**Review by kdmccormick — CHANGES_REQUESTED** [review]
I tried this out in a plugin of mine and hit a few issues, but overall I really like the interface. Being able to add proper Click commands that execute commands in containers feels like a huge step in capability for the plugin API.

File: tutor/commands/jobs.py
Comment (kdmccormick) [reply tutor/commands/jobs.py:62]: OK, I thought about this more.

I was happy to realize that everything in the new `initialise` function above could basically have also been written in a plugin. There calls to `read_template_file` and `echo_info`, but those are just simple wrappers around standard Python library functions. This is very cool, and it signals to me that `CLI_DO_COMMANDS` is a well-designed interface 👍🏻 

On the other hand, the naming & signature of this corner API feels inconsistent now:
* `CLI_DO_COMMANDS` returns Click commands that yield (service, command template) pairs
* `COMMANDS_INIT` returns (service, command template path) pairs.

The term "command" here is being overloaded between a few different concepts:
1. "command" as in "Click command"
2. "command" as in "a shell command to be run in a container"
3. "command" as in "a (service, shell command) tuple". In some places, we call this a *task*, which I like.

How would you feel about deprecating `COMMANDS_[PRE_]INIT` in favor of new filters:
```python
    #: List of shell commands to be executed during initialization. These commands typically
    #: include database migrations, setting feature flags, etc.
    #:
    #: :parameter list[tuple[str, str]] tasks: list of ``(service, command)`` tasks.
    #:
    #:     - ``service`` is the name of the container in which the task will be executed.
    #:     - ``command`` is a shell command template which will be rendered and executed in ``service``.
    CLI_DO_INIT_TASKS = filters.get("cli:commands:do:init")

    #: ...
    CLI_DO_PRE_INIT_TASKS = filters.get("cli:commands:do:init:pre")
```
(Names up for debate, of course. I wanted to choose names that made it clear that this filter is for adding *tasks* to the existing `do init` command.)

In addition to having different names, these filters have different signatures: they are expected to return `(service, command)` tuples instead of `(service, path)` tuples. This is congruent with the signature of `CLI_DO_COMMANDS`. The mapping from the old filters to the new ones would be pretty straightforward. For example:
```python
    with hooks.Contexts.APP("lms").enter():
        hooks.Filters.COMMANDS_INIT.add_item(("lms", ("hooks", "lms", "init")))
```
into:
```python
    with hooks.Contexts.APP("lms").enter():
        hooks.Filters.CLI_DO_INIT_TASKS.add_item(
            ("lms", env.read_template_file("hooks", "lms", "init"))
        )
```

I realize that this would be a breaking change, and would best be done in a different PR.

File: tutor/commands/jobs.py
Comment (regisb) [reply tutor/commands/jobs.py:37]: I think it's fine if we move the init scripts. It should not be a breaking change. I'll move the scripts to `tutor/templates/jobs/init/<service>.sh`.

File: tutor/commands/k8s.py
Comment (regisb) [reply tutor/commands/k8s.py:553]: Right. I confess that I did not even take the time to try out the API...

File: tutor/commands/compose.py
Comment (regisb) [reply tutor/commands/compose.py:304]: Right. It took me a very long time to figure out how to properly call `do` subcommands from quickstart, which is why I decided to get rid of the `--mount` option. But it's actually pretty simple to add it back, so I'll do that now.

File: tutor/commands/jobs.py
Comment (regisb) [reply tutor/commands/jobs.py:62]: Yes, your proposition makes a lot of sense. A couple precisions:

- Let's add the `CLI_DO_INIT_TASKS` filter.
- Let's keep the `COMMANDS_INIT` and `COMMANDS_PRE_INIT` filters for now. I don't want to force plugin maintainers to upgrade their plugins right away. We can just print a deprecation warning for the moment.
- We don't even need the `CLI_DO_PRE_INIT_TASKS` filter if we implement filter priorities. This is very easy, based on the existing action priority code.

I'll make these changes right now.

File: tutor/commands/jobs.py
Comment (kdmccormick) [inline tutor/commands/jobs.py:196]: Ugh, I just banged my head against my keyboard for half an hour before realizing why this wasn't working:
```python
def myjob():
    return [("lms", "echo hello world")]
hooks.Filters.CLI_DO_COMMANDS.add(myjob)
```
(can you see it?)
Other than the strongly typed hooks PR you have open, I'm not sure what else we would do to save a plugin developer from this kind of mistake.

**Review by kdmccormick — APPROVED** [review]
Looks good and works on my machine! The only suggestion I feel strongly about is the "jobs" vs "tasks" one.

File: tutor/jobs.py
Comment (regisb) [reply tutor/jobs.py:10]: I Agree.

File: tutor/commands/compose.py
Comment (regisb) [reply tutor/commands/compose.py]: `subcommand_metavar`  is great, but I'm not sure I want to go down that rabbit hole...

OK I just did :sweat_smile: Let's hope it's not making the code too complex.

File: tutor/commands/jobs.py
Comment (regisb) [reply tutor/commands/jobs.py:218]: `add_item`? To your credit, yes I must admit it's a little confusing. We'll see whether typed hooks help resolve this situation.

Comment (kdmccormick) [reply tutor/commands/jobs.py:218]: I think it will.

File: tutor/commands/compose.py
Comment (kdmccormick) [inline tutor/commands/compose.py:12]: Can we rename BaseJobContext -> BaseTaskContext ?

File: tutor/commands/k8s.py
Comment (kdmccormick) [inline tutor/commands/k8s.py:52]: We may want to address the mismatched meaning of "jobs" between Tutor and K8s:
```suggestion
class K8sTaskRunner(BaseTaskRunner):
    """
    Run tasks (bash commands) in Kubernetes-managed services.
    
    Terminology note:
        A Tutor "task" corresponds to a Kubernetes "job".
        In contrast, a "Tutor job" (invoked on the CLI using `tutor k8s do`)
        will trigger one or more Tutor tasks (Kubernetes jobs).
    """
```

File: tutor/commands/k8s.py
Comment (kdmccormick) [inline tutor/commands/k8s.py:349]: The rest of the command helptexts are written in a singular voice (e.g. `run  --  Run a command in a new container`) so I suggest doing the same here.
```suggestion
    Run a custom job in the right containers.
```

File: tutor/commands/compose.py
Comment (kdmccormick) [inline tutor/commands/compose.py:307]: Same suggestion.
```suggestion
    Run custom a job in the right containers.
```

File: tutor/commands/compose.py
Comment (kdmccormick) [inline tutor/commands/compose.py:81]: Didn't you mention on an issue or PR somewhere that we could phase out the `*.jobs.yml` compose files at some point? Was it contingent on some refactoring?

If not, I think `*.task.yml` and `COMPOSE_LOCAL_TASKS_TMP` would be a more appropriate names for the compose file and filter. I don't want to block this PR any longer, so I'm happy to make a backlog ticket for that if you agree.

**Review by kdmccormick — APPROVED** [review]
A few more suggestions, but again, incorporate whatever you like and merge as you see fit.

File: tutor/commands/k8s.py
Comment (regisb) [reply tutor/commands/k8s.py:52]: Good point, I improved the docs.

File: tutor/commands/compose.py
Comment (regisb) [reply tutor/commands/compose.py:81]: Not so much an issue, more of an open question that I raised here: https://github.com/overhangio/tutor/issues/642
I don't want to phase out or rename this template or the corresponding patches at this point, as it would break plugins -- I'd rather not break plugins just for a name change. The rename would affect the following files:

- tutor/templates/k8s/jobs.yml
- tutor/templates/dev/docker-compose.jobs.yml
- tutor/templates/local/docker-compose.jobs.yml

And patches:

- local-docker-compose-jobs-services
- dev-docker-compose-jobs-services
- k8s-jobs

I don't think that the name change warrants a new backlog ticket; "jobs" might not be the perfect name but it's close enough.

Comment (kdmccormick) [reply tutor/commands/compose.py:81]: Sounds good.

**Pull request merged**
